### PR TITLE
perf(layers): fast path for single-InputSpec assert_input_compatibility

### LINKED ISSUE #23275: assert_input_compatibility re-validates the same static InputSpec every call
`Layer.__call__` invokes `assert_input_compatibility()` (`keras/src/layers/input_spec.py`) on every single call. For most layers `self.input_spec` is a single, static `InputSpec` fixed once `build()` runs, but the function still unconditionally does the full generic-structure path every time: `tree.flatten(input_spec)`, `tree.flatten(inputs)`, a length check, and then per-index ndim/dtype/shape validation against the spec. None of that inputs-shape/dtype checking work changes between calls when the spec and the layer's expected rank/dtype are already fixed, so it's pure re-derivation of the same answer on every forward pass.

Fixed by adding a fast path for the common case: a single, non-nested `InputSpec` with a single tensor input, which skips straight to the scalar checks instead of going through the tree-flatten/list machinery.

Part of the per-call Python-dispatch overhead series in #22561. Fixed by #23184.

### PR DESCRIPTION
**Pull request by pctablet505**
Closes #23275

## Summary

`assert_input_compatibility` flattens both the spec and the inputs and runs a full input-count check before validating anything, even for the common case of one `InputSpec` against one tensor. This adds a fast path for that case and extracts the per-spec validation into `_check_input_spec`, which the general path also uses.

`_check_input_spec` is defined at the bottom of the file so the diff reads as a change to `assert_input_compatibility` plus a new helper, rather than a rewrite.

## Behaviour changes

Two, both intentional, neither a consequence of the fast path:

- An out-of-range `spec.axes` key now raises `ValueError` naming the axis, the input rank and the full shape. Previously it raised a bare `IndexError` from the tuple lookup.
- `allow_last_axis_squeeze` now squeezes only the side whose rank is one greater. Previously master squeezed a trailing `1` off **both** the input and the spec unconditionally, and then `zip` truncated the comparison to the shorter shape, so mismatches slipped through.

A rank mismatch that squeezing cannot resolve still does not raise here — `zip` truncates to the leading dimensions.

### This second change is user-visible, and wider than the two examples above

`Functional` sets `allow_last_axis_squeeze=True` on every input spec, so this tightens `model(x)` for real models. Please read this part before merging.

A model declared for grayscale accepts RGB on master and raises here:

```python
inp = keras.Input(shape=(28, 28, 1))
m = keras.Model(inp, keras.layers.Identity()(inp))
m(np.zeros((4, 28, 28, 3), dtype="float32"))
# master: returns, output shape (4, 28, 28, 3)
# this PR: ValueError, expected shape=(None, 28, 28, 1), found shape=(4, 28, 28, 3)
```

Scope, measured rather than estimated:

- **48 of 168** spec-shape x input-shape combinations flip from passing to `ValueError`, and **22 of 99** of those are *equal-rank* — i.e. unrelated to the rank-delta gating in the examples above.
- **14 of 151** probed real Functional model calls flip.
- It is **not confined to the new fast path.** The general path now calls the same `_check_input_spec`, so list-wrapped specs tighten identically. That is the point of sharing the helper, but it does widen the blast radius.

Every flip I could find is a genuine incompatibility: against an oracle of the documented `allow_last_axis_squeeze` semantics there were **0 violations**, and **0 cases anywhere** where this PR is more permissive than master. The `keras/src/layers` and `keras/src/models` failure set is unchanged. So master was silently accepting wrong-shaped input, and this is the fix — but it is a behaviour change in a PR framed as a perf refactor, and downstream code that was (wrongly) passing will now raise. Happy to split it out, gate it, or add a release note, whichever you prefer.

Error messages now report the shape the caller passed. Previously the squeeze rebound the local, so `(2, 3, 99, 1)` was reported as `found shape=(2, 3, 99)` — a shape the caller never used. Pinned by `test_error_reports_the_shape_that_was_passed`.

## Coverage

`input_spec_test.py` adds 27 tests. Twenty-two call `assert_input_compatibility` directly and cover both squeeze directions, the out-of-bounds axis, wrong axes, wrong shape, wrong dtype, `min_ndim`/`ndim`/`max_ndim`, non-tensor inputs, the optional-spec and empty-spec early returns, and that a bare spec and a one-element list spec behave identically — i.e. the fast path and the general path agree. The remaining five go through `Dense` and through `Functional` models, including the rank-standardization handoff described above.

## Contributor Agreement

- [x] This PR is linked to an issue that has been assigned to me (see `Closes #23275` above).
- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: keras/src/layers/input_spec.py [hunk 1]
```diff
@@ -119,6 +119,94 @@ def from_config(cls, config):
...
+                )
+    # Check shape.
+    if spec.shape is not None:
+        spec_shape = spec.shape
+        if spec.allow_last_axis_squeeze:
+            if shape and shape[-1] == 1:
+                shape = shape[:-1]
+            if spec_shape and spec_shape[-1] == 1:
+                spec_shape = spec_shape[:-1]
+        for spec_dim, dim in zip(spec_shape, shape):
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras/src/layers/input_spec.py:227]: Fixed the underlying bug (one-sided squeeze gated on rank-delta-exactly-one, so `(2, 3, 99, 1)` vs spec `(None, 3, 1)` and `(2, 1)` vs spec `(None, 3)` now correctly raise), but kept the rank-mismatch case that squeezing can't resolve as non-raising here, diverging from the suggested strict rank-equality check.

`allow_last_axis_squeeze=True` is set only by `Functional` (`keras/src/models/functional.py`), which calls `_adjust_input_rank` right after this check and raises its own `ValueError` there — one that includes the input's path in the nested structure, which this function doesn't have context for. That handoff is pinned by `functional_test.py::test_rank_standardization_failure` on master. A strict rank check here would also reject documented-legal inputs such as shape `(1,)` against spec `(None, 1)`.

Added regression tests for both squeeze directions and the out-of-bounds case in the same commit (4be28f4).

File: keras/src/layers/input_spec.py [hunk 2]
```diff
@@ -119,6 +119,94 @@ def from_config(cls, config):
...
+            if value is not None and shape[axis] not in {
+                value,
+                None,
+            }:
+                raise ValueError(
+                    f"Input {input_index} with name '{spec.name}' of layer "
+                    f"'{layer_name}' is incompatible with the layer: "
+                    f"expected axis {axis} of input shape to have value "
+                    f"{value}, but received input with shape {shape}"
+                )
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras/src/layers/input_spec.py:208]: Adopted — out-of-range `spec.axes` now raises a `ValueError` with the axis, ndim, and full shape instead of an unhandled `IndexError`. In 4be28f4.

File: keras/src/layers/input_spec_test.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+    def test_wrong_axes_raises(self):
+        spec = InputSpec(min_ndim=2, axes={-1: 3})
+        x = backend.convert_to_tensor(
+            np.array([[1.0, 2.0]])
+        )  # last dim=2 not 3
+        with self.assertRaisesRegex(ValueError, "axis"):
+            assert_input_compatibility(spec, x, "dense")
+
+    def test_wrong_shape_raises(self):
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras/src/layers/input_spec_test.py:120]: Added — `test_allow_last_axis_squeeze_rank_mismatch_raises` plus companions for the other squeeze direction and the out-of-bounds axis case, in 4be28f4.

File: keras/src/layers/input_spec.py [hunk 4]
```diff
@@ -119,6 +119,122 @@ def from_config(cls, config):
         return cls(**config)
 
 
+def _check_input_spec(spec, x, input_index, layer_name):
+    """Checks compatibility between a single input and a single `InputSpec`.
+
```
Comment (hertschuh) [inline keras/src/layers/input_spec.py:124]: This is mostly the same code as what was below in `assert_input_compatibility` but there are a number of non-trivial changes, which are really hard to review right now.

Can you move this function down at the bottom of the file so that I can see the diff?

File: keras/src/layers/input_spec.py [hunk 5]
```diff
@@ -139,6 +255,16 @@ def assert_input_compatibility(input_spec, inputs, layer_name):
     if not input_spec:
         return
 
+    # Fast path for the most common case: a single `InputSpec` and a single
+    # tensor input. This skips the `tree.flatten` calls and the input count
+    # check. Nested inputs (including any registered pytree) are detected by
+    # `tree.is_nested` and go through the general path below.
```
Comment (hertschuh) [inline keras/src/layers/input_spec.py:261]: Shorten this to:
```
    # Fast path for the most common case: a single `InputSpec` and a single
    # tensor input.
```

**Review by hertschuh — CHANGES_REQUESTED** [review]

**Comment by pctablet505** [comment]
Thanks for the review @hertschuh! All requested changes have been applied in `7a2a5d980`:

1. Moved `_check_input_spec` to the bottom of `keras/src/layers/input_spec.py` as a module-level private helper.
2. Shortened the `try / except AttributeError` comment to the exact 2-line wording requested.
3. Restored the original squeezed shape `(batch, ..., 1)` in the `allow_last_axis_squeeze` error message when rank mismatches after squeezing the last axis.
4. Added unit tests covering `allow_last_axis_squeeze` valid rank-mismatch inputs (e.g. `(4, 28, 28, 1)` into expected `(4, 28, 28)`) and error cases (`(4, 28, 28, 1)` into `(4, 32)`). Note that equal-rank inputs with a mismatched last axis (e.g. `(4, 28, 28, 3)` into expected `(None, 28, 28, 1)`) now correctly raise `ValueError` on axis `-1`, whereas previously `allow_last_axis_squeeze` only validated `axes` inside the `len(shape) - 1 == spec.ndim` branch.
